src/backend/distributed/connection/connection_management.c

```diff
@@ -438,186 +459,370 @@ ShutdownConnection(MultiConnection *connection)
 			}
 
-			pollResult = (select) (selectFileDescriptor + 1, &readFileDescriptorSet,
-								   &writeFileDescriptorSet, &exceptionFileDescriptorSet,
-								   &selectTimeout);
-#endif /* HAVE_POLL */
-
-			if (pollResult == 0)
+			if (event->events & WL_LATCH_SET)
 			{
-				/*
-				 * Timeout exceeded. Two things to do:
-				 * - check whether any interrupts arrived and handle them
-				 * - check whether establishment for connection already has
-				 *   lasted for too long, stop waiting if so.
-				 */
+				ResetLatch(MyLatch);
+
 				CHECK_FOR_INTERRUPTS();
 
-				if (TimestampDifferenceExceeds(connection->connectionStart,
-											   GetCurrentTimestamp(),
-											   NodeConnectionTimeout))
+				if (InterruptHoldoffCount > 0 && (QueryCancelPending || ProcDiePending))
 				{
-					ereport(WARNING, (errmsg("could not establish connection after %u ms",
-											 NodeConnectionTimeout)));
-
-					/* close connection, otherwise we take up resource on the other side */
-					PQfinish(connection->pgConn);
-					connection->pgConn = NULL;
-					break;
+					/*
+					 * because we can't break from 2 loops easily we need to not forget to
+					 * reset the memory context
+					 */
+					MemoryContextDelete(MemoryContextSwitchTo(oldContext));
+					return;
 				}
+
+				continue;
 			}
-			else if (pollResult > 0)
-			{
-				/*
-				 * IO possible, continue connection establishment. We could
-				 * check for timeouts here as well, but if there's progress
-				 * there seems little point.
-				 */
-				break;
-			}
-			else
+
+			connectionStateChanged = MultiConnectionStatePoll(connectionState);
+			if (connectionStateChanged)
 			{
-				int errorCode = errno;
-#ifdef WIN32
-				errorCode = WSAGetLastError();
-#endif
-				if (errorCode == EINTR)
+				if (connectionState->phase != MULTI_CONNECTION_PHASE_CONNECTING)
 				{
-					/* Retrying, signal interrupted. So check. */
-					CHECK_FOR_INTERRUPTS();
+					/* we cannot stop waiting for connection, so rebuild the event set */
+					waitEventSetRebuild = true;
 				}
 				else
 				{
-					/*
-					 * We ERROR here, instead of just returning a failed
-					 * connection, because this shouldn't happen, and indicates a
-					 * programming error somewhere, not a network etc. issue.
-					 */
-					ereport(ERROR, (errcode_for_socket_access(),
-									errmsg("poll()/select() failed: %m")));
+					/* connection state changed, reset the event mask */
+					uint32 eventMask = MultiConnectionStateEventMask(connectionState);
+					ModifyWaitEvent(waitEventSet, event->pos, eventMask, NULL);
 				}
 			}
 		}
+
+		if (eventCount == 0)
+		{
+			/*
+			 * timeout has occured on waitset, double check the timeout since
+			 * connectionStart and if passed close all non-finished connections
+			 */
+
+			TimestampTz now = GetCurrentTimestamp();
+			if (TimestampDifferenceExceeds(connectionStart, now, NodeConnectionTimeout))
+			{
+				CloseNotReadyMultiConnectionStates(connectionStates);
```

Comment:
I kind of looked over this before. Why do we need it? We close all connections at transaction end in `AfterXactConnectionHandling`.

Reply:
This is ported from the original code around this comment https://github.com/citusdata/citus/pull/2584/files#diff-0b04fef38ab380718daadfcf21d40eb0L579
```
/* close connection, otherwise we take up resource on the other side */
```

I did not look deeply into the reasoning, what you are suggesting is that the comment is not correct? Given the claim that it would be closed in `AfterXactConnectionHandling`.

Reply:
Hmz, there are places where we try to be tolerant to connection failures, for example when querying a reference table in the router executor. If we time out on the first replica we will then fail over to the second replica. There are also particular functions that open many connections, but can tolerate some of them failing (e.g. `get_global_active_transactions`).

If we continue with the transaction after a timeout it makes sense to close the connection.

We have various functions that allow caller to specify whether it can tolerate ERRORs by passing a boolean flag (`raiseErrors`). What if we do that here as well? When in doubt, we can default to false since that's the existing behaviour.

Reply:
I am not fond of adding the `raiseErrors`, would suggest to use `PG_TRY` and `PG_CATCH` if you can tolerate ERRORs.

On the other hand, the reason I changed the code to raise an error is to be able to call CloseConnection instead of PQfinish on the connection. From your previous comment and testing of a `DDL` and `Router query` I was able to confirm that the connection got closed via `AfterXactConnectionHandling`. That would suggest we can change the message back to a warning and have the cases described above still behave the same.

I will try and add tests for the timeout on reference table query and the `get_global_active_transactions` given tests for those are most likely missing in the failure tests. Given your description it sounds like essential tests to have to make this change confidently.

Reply:
In general, PG_TRY and PG_CATCH should be used sparingly because weird things happen to your variables when you jump through C code https://linux.die.net/man/3/longjmp . 

Changing back to a warning seems fine.